### ISSUE TITLE
Rom loading from menu fails

### ISSUE THREAD

**Issue by rattboi** [DragonMinded/netboot#7]
Running: 

`./netdimm_menu --verbose 192.168.0.2 ../roms/`

Rom list appears, I select a game, and get the following:

```
...
Attaching section to a new ROM section at the end of the file.
Connecting to net dimm...
Sending menu to net dimm...
length: 0017168c
Talking to net dimm to wait for ROM selection...
Received type: 0x1000, length: 4
Requested MetalSlug6 be loaded...
Sending type: 0x1009, length: 8
Packet transfer took 0.28948378562927246 seconds
Could not send data to NetDimm
Sending game failed...
```

**Comment by DragonMinded** [comment]
What platform? I assume its hitting the same OSX bug as the other open issue?

**Comment by rattboi** [comment]
Not OSX. This is running from rpi.

**Comment by rattboi** [comment]
I'm not getting any exceptions in `./netdimm_menu`, it's just failing to send

**Comment by DragonMinded** [comment]
It swallows exceptions on purpose, stick a print(str(e)) in the exception handler to see what it's whining about.

**Comment by DragonMinded** [comment]
Or run with --verbose, I think that gives you a bit more info?

**Comment by rattboi** [comment]
haha, of course it worked for the first time ever when I threw the `print(str(e))` in `__send_packet`. I'll rerun a few times to try and capture the exception.

**Comment by DragonMinded** [comment]
Send packet isn't going to have anything, that's just for messaging. The netdimm module will instead.

**Comment by rattboi** [comment]
Oh, indeed, this does look similar to the OSX problem. I'm getting occasional ```[Errno 32] Broken pipe``` errors

```
Sending menu to net dimm...
length: 0017168c
Talking to net dimm to wait for ROM selection...
Received type: 0x1000, length: 4
Requested Ikaruga v3 be loaded...
Sending type: 0x1009, length: 8
Packet transfer took 0.2052931785583496 seconds
[Errno 32] Broken pipe
Could not send data to NetDimm
Sending game failed...
```

**Comment by DragonMinded** [comment]
Bizarre, I've never encountered that on windows/linux, only heard reports from OSX folks. What distro do you have on your pi?

**Comment by DragonMinded** [comment]
I have a sneaking suspicion that we can catch broken pipe in the netdimm module, retry the connection and resend what we wanted to send and it should work.

**Comment by rattboi** [comment]
I'm just running raspbian. I did just try something based on the OSX report and I think maybe it's related?

I tried putting a `time.sleep(20)` in just before doing `send_message(MESSAGE_LOAD_PROGRESS)` just to give the netdimm a bunch of time to "catch up", and what I saw was that the menu-side errored with a "communication error", way earlier than the 20 second sleep. I suppose this is the "10 second timeout" you were referring to in #6 . If so, I highly doubt that is actually 10 seconds on the menu side. It seemed closer to 2-3 seconds. I'll try again and report back.

**Comment by DragonMinded** [comment]
Yeah, the menu-side of things expects to be communicated with constantly to detect failures.

**Comment by DragonMinded** [comment]
The net dimm shouldn't need to catch up in that section, given that it isn't rebooting. That's one advantage of sending before a reboot instead of displaying a now loading screen.

**Comment by rattboi** [comment]
I've continued to poke at this, and have some small tidbits to report:

```
Talking to net dimm to wait for ROM selection...
Received type: 0x1000, length: 4                      
Requested CosmicSmash be loaded...                                   
Sending type: 0x1009, length: 8               
Packet transfer took 0.21047401428222656 seconds     
after load progress                              
  File "./netdimm_menu", line 11, in <module>                
    runpy.run_module(f"scripts.{name}", run_name="__main__")                
  File "/usr/lib/python3.7/runpy.py", line 208, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code 
    exec(code, run_globals)                            
  File "/home/pi/netboot/scripts/netdimm_menu.py", line 766, in <module>
    sys.exit(main())                                     
  File "/home/pi/netboot/scripts/netdimm_menu.py", line 697, in main
    netdimm.send(selected_file, disable_crc_check=True, disable_now_loading=True, progress_callback=progress_callback)
  File "/home/pi/netboot/netdimm/netdimm.py", line 157, in send
    self.__upload_file(data, key, progress_callback or (lambda _cur, _tot: None))
  File "/home/pi/netboot/netdimm/netdimm.py", line 611, in __upload_file
    self.__upload(sequence, addr, current, last_packet)
  File "/home/pi/netboot/netdimm/netdimm.py", line 444, in __upload
    self.__send_packet(NetDimmPacket(0x04, 0x81 if last_chunk else 0x80, struct.pack("<IIH", sequence, addr, 0) + data))
  File "/home/pi/netboot/netdimm/netdimm.py", line 295, in __send_packet
    traceback.print_stack()                     
Could not send data to NetDimm                    
Sending game failed, will try again...                
Waiting for cabinet to be ready to receive the menu...        
```

I've added `--persistent` to the menu call, and 9/10 times, the game upload is failing. Peculiarly, the menu reload after this works basically every time.

**Comment by rattboi** [comment]
To further narrow this down, I can move from rpi -> my linux PC and see if it continues. Is there any chance it's related to the naomi side of things re: netdimm version, etc?

**Comment by DragonMinded** [comment]
It might be, but I've tested this on 3.06, 3.17, 4.01 and 4.02 to success on all of them. What's your dimm version?

**Comment by rattboi** [comment]
3.17

**Comment by DragonMinded** [comment]
Long time, but we just got a fix for the OSX-side of things from @ArcadeHustle and given the similarity to this bug it might fix your case too. Check out this line of code for escape hatches: https://github.com/DragonMinded/netboot/blob/trunk/netdimm/netdimm.py#L301

tl;dr - Run your stuff with "ALTERNATE_TIMEOUT_HANDLING=1" environment variable set and see if it fixes the problem.

**Comment by rattboi** [comment]
Thanks so much! I'll try to test it soon.

**Comment by DragonMinded** [comment]
Alternatively, can you just grab the latest with my timeout bump and test that without the environment flag set? I think that should be good enough for raspibian.

**Comment by DragonMinded** [comment]
Any luck?

**Comment by rattboi** [comment]
Short answer is no, but not because I didn't try. I seemingly bent a few pins on my Naomi and now the netdimm won't start! I'm trying to repair it, but it is sensitive, as you might expect.

**Comment by DragonMinded** [comment]
OH NO, does it respond to network requests at least or is it totally dead?

**Comment by rattboi** [comment]
I believe it is fixable. I just need to take some time to do so..

**Comment by DragonMinded** [comment]
Again poking this.

**Comment by rattboi** [comment]
I haven't had the time, but my school quarter is over tomorrow, so I *should* have some time to try and repair my naomi in the next week or so. I'll keep you posted. I'm also excited to see if the recent fixes fix my issues.

**Comment by DragonMinded** [comment]
Another friendly ping, were you able to test?

**Comment by rattboi** [comment]
Got it working again! I tried both with and without the flag and no dice yet. I'll set up to use a Linux pc instead to A/B test and get back with the results.

**Comment by rattboi** [comment]
Just loaded a game successfully from my laptop. So we know it *can* work with my Naomi. I'll start over on the rpi fresh and see if I get any further.

**Comment by rattboi** [comment]
With a clean checkout, I've successfully been able to load reliably when using ALTERNATE_TIMEOUT_HANDLING. 🎉

**Comment by ArcadeHustle** [comment]
2 birds with one stone!

**Comment by DragonMinded** [comment]
Sounds like this can be closed then!
